File: javascript/quick-sort.js
function quickSort(arr) {
    if (arr.length < 2) return arr;

    // take out first index for pivot
    let pivot = arr.shift();
    let arrLen = arr.length;
    let left = [];
    let right = [];

    for(let itr= 0; itr < arrLen; itr++) {
        // distribute to left part if arr equal or smaller than pivot
        if (arr[itr] <= pivot) {
            left.push(arr[itr]);
        }
        // for the remaining, put to right side. it must be greater than pivot.
        else if (arr[itr] > pivot) {
            right.push(arr[itr])
        }
    }

    // put them together with pivot in center. left - pivot - right
    return [].concat(quickSort(left), [pivot], quickSort(right));
}

console.log(quickSort([4,3,2,1,5,6,7,9,8]));

File: golang/quick-sort/quick-sort.go
package main

import "fmt"

func quickSort(arr []int) []int {
	if len(arr) < 2 {
		return arr
	}
	// take out first index for pivot
	pivot := arr[0]
	arrData := arr[1:]
	arrLen := len(arrData)
	var left, right []int
	for idx := 0; idx < arrLen; idx++ {
		if arrData[idx] <= pivot { // distribute to left part if arr equal or smaller than pivot
			left = append(left, arrData[idx])
		} else if arrData[idx] > pivot { // for the remaining, put to right side. it must be greater than pivot.
			right = append(right, arrData[idx])
		}
	}

	merge := append(quickSort(left), pivot)
	merge = append(merge, quickSort(right)...)
	return merge
}

func main() {
	result := quickSort([]int{5, 4, 3, 2, 1})
	fmt.Println("result", result)
}
